### DISCUSSION TITLE
How to strictify ZodType

### DISCUSSION THREAD

**Discussion by Rigo-m** [colinhacks/zod#2784] [Q&A] [answered]
Using z.ZodType<Obj>, typescript tells me that I'm missing the name key from the objSchema, correctly.

```ts
type Obj = {
  id: string
  name: string
}

const objSchema: z.ZodType<Obj> = z.object({
  id: z.string()
}).strict() 
```

But, this example:

```ts
type Obj = {
  id: string
  name: string
}

const objSchema: z.ZodType<Obj> = z.object({
  id: z.string(),
  name: z.string(),
  extraParam: z.string() // unwanted extra param
}).strict()
```

Doesn't warn me that I have an extra key in my schema definition. How can I achieve the same level of strictness from the generic type of the strict() method?

**Comment by conorbrandon** [comment, marked as answer]
I believe this is a byproduct of TS's [structural typing](https://www.typescriptlang.org/play#example/structural-typing). For example, [this](https://www.typescriptlang.org/play?#code/MYGwhgzhAECiAeYC2AHEBTAPAFWu+ALugHYAmMASusAPYBOpmEBdAlsQOYA00zbnAPgHQA3gChovABZgU6AFzRsAbgnQUbAG5gi0WsT4BXYAXoAKCDLmLsASlFrJBKawgA6S7PTQAvNK+qkgC+asw6rMB6dOg6WACqeIQk5NBUtAxMLOzcvFmCAhZWCtBx9uKSktEEhnTE0MToAO5wiKgYhV62gdAhIfrMiYoIyGhYIgBmNDSKfNlBwn7DbehuwNGxZhNTigBEOzwARmB0uztBXWJAA) code exhibits the same behavior:

```ts
class Example<T extends Record<string, string>> {
  shape: T;
  private constructor(shape: T) {
    this.shape = shape;
  }
  static create<U extends Record<string, string>>(shape: U) {
    return new Example(shape);
  }
}
const ex: Example<{foo: string}> = Example.create({foo: "", bar: ""});
```
